### PR TITLE
Use singleflight for caching environment.go calls

### PR DESCRIPTION
**Pull request by oliveromahony**
### Proposed changes

Use singleflight for the some of the environment calls in environment.go. This should speed up performance.

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [`CONTRIBUTING`](https://github.com/nginx/agent/blob/main/docs/CONTRIBUTING.md) document
- [x] I have run ```make install-tools``` and have attached any dependency changes to this pull request
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works
- [x] If applicable, I have checked that any relevant tests pass after adding my changes
- [ ] If applicable, I have updated any relevant documentation ([`README.md`](https://github.com/nginx/agent/blob/main/README.md))
- [ ] If applicable, I have tested my cross-platform changes on Ubuntu 22, Redhat 8, SUSE 15 and FreeBSD 13

### CODE REVIEW

File: src/core/environment.go [hunk 1]
```diff
@@ -185,35 +189,36 @@ func getUnixName() string {
...
+		hostname = ""
 	}
-	return hostInformation.Hostname
+	return hostname
 }
 
 func (env *EnvironmentType) GetSystemUUID() string {
-	ctx := context.Background()
-	defer ctx.Done()
+	res, _, _ := singleflightGroup.Do(GetSystemUUIDKey, func() (interface{}, error) {
```
Comment (dhurley) [inline src/core/environment.go:201]: `singleflightGroup.Do` (https://pkg.go.dev/golang.org/x/sync/singleflight#Group.Do) can return an error. Can you add a check for that?

**Review by dhurley — COMMENTED** [review]
Can you also update the title of the PR?

File: src/core/environment.go [hunk 2]
```diff
@@ -307,18 +334,21 @@ func (env *EnvironmentType) IsContainer() bool {
 		k8sServiceAcct = "/var/run/secrets/kubernetes.io/serviceaccount"
 	)
 
-	for _, filename := range []string{dockerEnv, containerEnv, k8sServiceAcct} {
-		if _, err := os.Stat(filename); err == nil {
-			log.Debugf("is a container because (%s) exists", filename)
-			return true
+	res, _, _ := singleflightGroup.Do(IsContainerKey, func() (interface{}, error) {
```
Comment (dhurley) [inline src/core/environment.go:337]: Can you handle the error returned here as well?

**Review by dhurley — APPROVED** [review]

File: src/core/environment.go [hunk 3]
```diff
@@ -194,24 +213,32 @@ func (env *EnvironmentType) GetHostname() string {
...
+		hostID, err := host.HostIDWithContext(ctx)
+		if err != nil {
+			log.Infof("Unable to read host id from dataplane, defaulting value. Error: %v", err)
+			return "", err
+		}
+		return uuid.NewMD5(uuid.Nil, []byte(hostID)).String(), err
+	})
 	if err != nil {
-		log.Infof("Unable to read host id from dataplane, defaulting value. Error: %v", err)
+		log.Warnf("unable to set hostname due to %v", err)
```
Comment (Dean-Coakley) [inline src/core/environment.go:237]: (Optionally follow logging convention)
```suggestion
		log.Warnf("Unable to set hostname due to: %v", err)
```

File: src/core/environment.go [hunk 4]
```diff
@@ -307,18 +334,25 @@ func (env *EnvironmentType) IsContainer() bool {
...
-		return result
+		// v1 check
+		if result, err := cGroupV1Check(selfCgroup); err == nil && result {
+			return result, err
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		log.Warnf("unable to retrieve values from cache (%v)", err)
```
Comment (Dean-Coakley) [inline src/core/environment.go:352]: (Optionally follow logging convention)
```suggestion
		log.Warnf("Unable to retrieve values from cache (%v)", err)
```

**Review by Dean-Coakley — APPROVED** [review]

**Review by aphralG — APPROVED** [review]

**Pull request merged**